File: backend/src/services/healthService.ts
import cron from 'node-cron';
import axios from 'axios';
import { prisma } from '../prisma';
import { Agent } from '@prisma/client';

const TIMEOUT_MS = parseInt(process.env.HEALTH_CHECK_TIMEOUT_MS || '5000');

// Runtime env overrides — checked first, falling back to DB value
const HEALTH_URL_OVERRIDES: Record<string, string | undefined> = {
  nexus: process.env.AGENT_NEXUS_HEALTH_URL,
  cipher: process.env.AGENT_CIPHER_HEALTH_URL,
  relay: process.env.AGENT_RELAY_HEALTH_URL,
};

function resolveHealthEndpoint(agent: Agent): string {
  const key = agent.name.toLowerCase();
  return HEALTH_URL_OVERRIDES[key] || agent.healthEndpoint;
}

async function checkAgentHealth(agent: Agent): Promise<boolean> {
  const url = resolveHealthEndpoint(agent);
  try {
    const response = await axios.get(url, { timeout: TIMEOUT_MS });
    return response.status >= 200 && response.status < 300;
  } catch {
    return false;
  }
}

async function runHealthCycle(): Promise<void> {
  try {
    const agents = await prisma.agent.findMany();
    let online = 0;

    for (const agent of agents) {
      const resolvedEndpoint = resolveHealthEndpoint(agent);
      const isOnline = await checkAgentHealth(agent);
      if (isOnline) online++;
      await prisma.agent.update({
        where: { id: agent.id },
        data: {
          isOnline,
          lastCheckedAt: new Date(),
          // Keep DB in sync with env overrides so API responses show the real URL
          ...(resolvedEndpoint !== agent.healthEndpoint && {
            healthEndpoint: resolvedEndpoint,
          }),
        },
      });
    }

    console.log(
      `[Health] Cycle complete. ${agents.length} checked. ${online} online. ${agents.length - online} offline. ${new Date().toISOString()}`
    );
  } catch (err) {
    console.error('[Health] Cycle error:', err);
  }
}

export async function triggerHealthCheck(): Promise<{
  checked: number;
  online: number;
  offline: number;
}> {
  const agents = await prisma.agent.findMany();
  let online = 0;

  for (const agent of agents) {
    const resolvedEndpoint = resolveHealthEndpoint(agent);
    const isOnline = await checkAgentHealth(agent);
    if (isOnline) online++;
    await prisma.agent.update({
      where: { id: agent.id },
      data: {
        isOnline,
        lastCheckedAt: new Date(),
        ...(resolvedEndpoint !== agent.healthEndpoint && {
          healthEndpoint: resolvedEndpoint,
        }),
      },
    });
  }

  return { checked: agents.length, online, offline: agents.length - online };
}

export function startHealthScheduler(): void {
  runHealthCycle();
  cron.schedule('*/5 * * * *', runHealthCycle);
  console.log('[Health] Scheduler started. Interval: 5 minutes');
}

File: backend/services/health_service.py
import asyncio
import logging
from datetime import datetime, timezone

import httpx
from apscheduler.schedulers.asyncio import AsyncIOScheduler

from config import settings

logger = logging.getLogger(__name__)
_scheduler: AsyncIOScheduler | None = None


async def run_health_checks():
    from models.agent import Agent

    agents = await Agent.find_all().to_list()
    checked = online = offline = 0
    timeout = settings.health_check_timeout_ms / 1000

    async with httpx.AsyncClient(timeout=timeout) as client:
        for agent in agents:
            checked += 1
            try:
                resp = await client.get(agent.health_endpoint)
                agent.is_online = resp.is_success
            except Exception:
                agent.is_online = False

            agent.last_checked_at = datetime.now(timezone.utc)
            await agent.save()

            if agent.is_online:
                online += 1
            else:
                offline += 1

    logger.info(
        "Health check complete — checked: %d, online: %d, offline: %d at %s",
        checked,
        online,
        offline,
        datetime.now(timezone.utc).isoformat(),
    )
    return {"checked": checked, "online": online, "offline": offline}


def start_health_scheduler():
    global _scheduler
    interval_seconds = settings.health_check_interval_ms / 1000
    _scheduler = AsyncIOScheduler()
    _scheduler.add_job(run_health_checks, "interval", seconds=interval_seconds, next_run_time=datetime.now(timezone.utc))
    _scheduler.start()
    logger.info("Health scheduler started (interval: %ds)", interval_seconds)


def stop_health_scheduler():
    global _scheduler
    if _scheduler is not None:
        _scheduler.shutdown(wait=False)
